File: sourcegen-generator-java/src/test/java/io/micronaut/sourcegen/javapoet/write/ConversionProgramRegressionTest.java
/*
 * Copyright 2017-2026 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.sourcegen.javapoet.write;

import io.micronaut.sourcegen.model.ClassDef;
import io.micronaut.sourcegen.model.ClassTypeDef;
import io.micronaut.sourcegen.model.ExpressionDef;
import io.micronaut.sourcegen.model.MethodDef;
import io.micronaut.sourcegen.model.StatementDef;
import io.micronaut.sourcegen.model.TypeDef;
import org.junit.jupiter.api.Test;

import javax.lang.model.element.Modifier;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static io.micronaut.sourcegen.javapoet.write.JavaCompileAssertions.compileMatchingSnapshots;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertSame;

/**
 * Executes DSL programs at conversion and scope boundaries.
 *
 * @since 2.3
 */
public class ConversionProgramRegressionTest {

    private static final String SNAPSHOTS = "/conversion-programs/java";

    @ParameterizedTest(name = "the array helper keeps the caller variable of kind {0}")
    @ValueSource(strings = {"T", "R", "method"})
    void arrayHelperPreservesCallerVariables(String kind) throws Exception {
        var callerVariable = TypeDef.variable(kind.equals("method") ? "R" : kind);
        var x = TypeDef.variable("X");
        var u = TypeDef.variable("U", TypeDef.parameterized(ClassTypeDef.of(List.class), x), TypeDef.of(Serializable.class));
        var arrays = MethodDef.builder("arrays").addModifiers(Modifier.PUBLIC).addTypeVariable(u)
            .addParameter("value", u.array()).returns(Object.class).build();
        var caller = ClassDef.builder("test.ArrayScope" + kind).addModifiers(Modifier.PUBLIC);
        var call = MethodDef.builder("call").addModifiers(Modifier.PUBLIC);
        if (kind.equals("method")) {
            call.addTypeVariable(callerVariable);
        } else {
            caller.addTypeVariable(callerVariable);
        }
        var def = caller.addMethod(call.addParameter("target", TypeDef.parameterized(ClassTypeDef.of(IntersectionTarget.class), callerVariable))
            .addParameter("value", Object.class).returns(Object.class)
            .build((self, p) -> p.getFirst().invoke(arrays, p.get(1)).returning())).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            var cls = loader.loadClass(def.getName());
            var value = new ArrayList<?>[]{new ArrayList<>(List.of("text"))};
            assertSame(value, cls.getMethod("call", IntersectionTarget.class, Object.class)
                .invoke(cls.getConstructor().newInstance(), new IntersectionTarget<String>(), value));
        }
    }

    @ParameterizedTest(name = "an inherited {0} call matches the full generic signature")
    @ValueSource(strings = {"overload", "array"})
    void inheritedCallsMatchTheFullGenericSignature(String kind) throws Exception {
        var x = TypeDef.variable("X", TypeDef.of(CharSequence.class));
        var u = TypeDef.variable("U", x);
        var parameter = kind.equals("array") ? TypeDef.variable("U").array() : u;
        var method = MethodDef.builder(kind.equals("array") ? "arrays" : "identity").addModifiers(Modifier.PUBLIC)
            .addTypeVariable(u).addParameter("value", parameter).returns(Object.class).build();
        var receiver = kind.equals("array") ? ArrayChild.class : GenericOverloadedChild.class;
        var def = ClassDef.builder("test.GenericSignature" + kind).addModifiers(Modifier.PUBLIC)
            .addMethod(MethodDef.builder("call").addModifiers(Modifier.PUBLIC).addParameter("target", receiver)
                .addParameter("value", Object.class).returns(Object.class)
                .build((self, p) -> p.getFirst().invoke(method, p.get(1)).returning())).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            var cls = loader.loadClass(def.getName());
            Object value = kind.equals("array") ? new String[]{"text"} : "text";
            assertSame(value, cls.getMethod("call", receiver, Object.class)
                .invoke(cls.getConstructor().newInstance(), receiver.getConstructor().newInstance(), value));
        }
    }

    @ParameterizedTest(name = "the bounds of a raw receiver do not capture the caller variable {0}")
    @ValueSource(strings = {"X", "Y"})
    void rawReceiverBoundsDoNotCaptureCallerVariables(String name) throws Exception {
        var x = TypeDef.variable("X", TypeDef.of(CharSequence.class));
        var u = TypeDef.variable("U", x);
        var identity = MethodDef.builder("identity").addModifiers(Modifier.PUBLIC).addTypeVariable(u)
            .addParameter("value", u).returns(Object.class).build();
        var def = ClassDef.builder("test.RawReceiver" + name).addModifiers(Modifier.PUBLIC).addTypeVariable(TypeDef.variable(name))
            .addMethod(MethodDef.builder("call").addModifiers(Modifier.PUBLIC).addParameter("target", CalleeBoundsRegressionTest.Parent.class)
                .addParameter("value", Object.class).returns(Object.class)
                .build((self, p) -> p.getFirst().invoke(identity, p.get(1)).returning())).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            var cls = loader.loadClass(def.getName());
            assertEquals("text", cls.getMethod("call", CalleeBoundsRegressionTest.Parent.class, Object.class)
                .invoke(cls.getConstructor().newInstance(), new CalleeBoundsRegressionTest.Parent<String>(), "text"));
        }
    }

    @ParameterizedTest(name = "a loop over a {0} constant condition completes as javac decides")
    @ValueSource(strings = {"primitive", "boxed", "negated", "comparison"})
    void loopCompletionMatchesJavaConstantExpressionRules(String kind) throws Exception {
        var condition = switch (kind) {
            case "boxed" -> ExpressionDef.trueValue().cast(Boolean.class).isTrue();
            case "negated" -> ExpressionDef.falseValue().isFalse();
            case "comparison" -> ExpressionDef.constant(1).compare(ExpressionDef.ComparisonOperation.OpType.EQUAL_TO, ExpressionDef.constant(1));
            default -> ExpressionDef.trueValue().isTrue();
        };
        var def = ClassDef.builder("test.ConstantLoop" + kind).addModifiers(Modifier.PUBLIC)
            .addMethod(MethodDef.builder("call").addModifiers(Modifier.PUBLIC).returns(int.class)
                .build((self, p) -> StatementDef.multi(condition.whileLoop(ExpressionDef.constant(1).returning()), ExpressionDef.constant(2).returning()))).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            var cls = loader.loadClass(def.getName());
            assertEquals(1, cls.getMethod("call").invoke(cls.getConstructor().newInstance()));
        }
    }

    @Test
    void adaptedReferenceCanReturnRawGeneratedGenericType() throws Exception {
        var box = ClassDef.builder("test.Box").addModifiers(Modifier.PUBLIC).addTypeVariable(TypeDef.variable("E")).build();
        var get = MethodDef.builder("get").addModifiers(Modifier.PUBLIC).overrides().returns(Object.class)
            .build((self, p) -> ExpressionDef.nullValue().returning());
        var target = ClassDef.builder("test.BoxSource").addModifiers(Modifier.PUBLIC)
            .addSuperinterface(TypeDef.parameterized(ClassTypeDef.of(Supplier.class), TypeDef.parameterized(ClassTypeDef.of(box), TypeDef.STRING)))
            .addMethod(get).build();
        var supplier = TypeDef.parameterized(ClassTypeDef.of(Supplier.class), TypeDef.parameterized(ClassTypeDef.of(box), TypeDef.OBJECT));
        var def = ClassDef.builder("test.BoxReferenceCall").addModifiers(Modifier.PUBLIC)
            .addMethod(MethodDef.builder("reference").addModifiers(Modifier.PUBLIC).addParameter("target", target.asTypeDef()).returns(supplier)
                .build((self, p) -> supplier.methodReference(p.getFirst(), get).returning())).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, box, target, def)) {
            var cls = loader.loadClass(def.getName());
            var targetClass = loader.loadClass(target.getName());
            var reference = (Supplier<?>) cls.getMethod("reference", targetClass)
                .invoke(cls.getConstructor().newInstance(), targetClass.getConstructor().newInstance());
            assertNull(reference.get());
        }
    }

    /**
     * Supplies an array parameter with a parameterized intersection bound.
     * @param <X> The element type
     * @since 2.3
     */
    public static class IntersectionTarget<X> {
        /**
         * @param value The array
         * @param <U> The component type
         * @return The same array
         */
        public <U extends List<X> & Serializable> Object arrays(U[] value) {
            return value;
        }
    }

    /**
     * Adds an unrelated generic overload to an inherited generic method.
     * @since 2.3
     */
    public static class GenericOverloadedChild extends CalleeBoundsRegressionTest.Parent<String> {
        /**
         * @param value The number
         * @param <U> The number type
         * @return The same number
         */
        public <U extends Number> U identity(U value) {
            return value;
        }
    }

    /**
     * Supplies an inherited generic array method.
     * @param <X> The component bound
     * @since 2.3
     */
    public static class ArrayParent<X extends CharSequence> {
        /**
         * @param value The array
         * @param <U> The component type
         * @return The same array
         */
        public <U extends X> Object arrays(U[] value) {
            return value;
        }
    }

    /**
     * Specializes the inherited array component bound.
     * @since 2.3
     */
    public static class ArrayChild extends ArrayParent<String> {
    }
}

File: sourcegen-generator-kotlin/src/test/kotlin/io/micronaut/sourcegen/ConversionProgramRegressionTest.kt
/*
 * Copyright 2017-2026 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.sourcegen

import io.micronaut.sourcegen.KotlinCompileAssertions.compileMatchingSnapshots
import io.micronaut.sourcegen.model.ClassDef
import io.micronaut.sourcegen.model.ClassTypeDef
import io.micronaut.sourcegen.model.ExpressionDef
import io.micronaut.sourcegen.model.MethodDef
import io.micronaut.sourcegen.model.StatementDef
import io.micronaut.sourcegen.model.TypeDef
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertSame
import java.io.Serializable
import javax.lang.model.element.Modifier
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

/** Executes DSL programs at conversion and scope boundaries. */
class ConversionProgramRegressionTest {

    private val snapshots = "/conversion-programs/kotlin"

    @ParameterizedTest(name = "the array helper keeps the caller variable {0}")
    @ValueSource(strings = ["T", "R"])
    fun arrayHelperPreservesCallerVariables(name: String) {
        val callerVariable = TypeDef.variable(name)
        val x = TypeDef.variable("X")
        val u = TypeDef.variable("U", TypeDef.parameterized(ClassTypeDef.of(List::class.java), x), TypeDef.of(Serializable::class.java))
        val arrays = MethodDef.builder("arrays").addModifiers(Modifier.PUBLIC).addTypeVariable(u)
            .addParameter("value", u.array()).returns(TypeDef.OBJECT).build()
        val def = ClassDef.builder("test.ArrayScope$name").addTypeVariable(callerVariable)
            .addMethod(MethodDef.builder("call").addParameter("target", TypeDef.parameterized(ClassTypeDef.of(IntersectionTarget::class.java), callerVariable))
                .addParameter("value", TypeDef.OBJECT).returns(TypeDef.OBJECT)
                .build { _, p -> p[0].invoke(arrays, p[1]).returning() }).build()
        compileMatchingSnapshots(snapshots, def).use { loader ->
            val cls = loader.loadClass(def.name)
            val value = arrayOf(arrayListOf("text"))
            assertSame(value, cls.getMethod("call", IntersectionTarget::class.java, Any::class.java)
                .invoke(cls.getConstructor().newInstance(), IntersectionTarget<String>(), value))
        }
    }

    @ParameterizedTest(name = "an inherited {0} call matches the full generic signature")
    @ValueSource(strings = ["overload", "array"])
    fun inheritedCallsMatchTheFullGenericSignature(kind: String) {
        val x = TypeDef.variable("X", TypeDef.of(CharSequence::class.java))
        val u = TypeDef.variable("U", x)
        val parameter = if (kind == "array") TypeDef.variable("U").array() else u
        val method = MethodDef.builder(if (kind == "array") "arrays" else "identity").addModifiers(Modifier.PUBLIC)
            .addTypeVariable(u).addParameter("value", parameter).returns(TypeDef.OBJECT).build()
        val receiver = if (kind == "array") ArrayChild::class.java else GenericOverloadedChild::class.java
        val def = ClassDef.builder("test.GenericSignature$kind")
            .addMethod(MethodDef.builder("call").addParameter("target", receiver).addParameter("value", TypeDef.OBJECT).returns(TypeDef.OBJECT)
                .build { _, p -> p[0].invoke(method, p[1]).returning() }).build()
        compileMatchingSnapshots(snapshots, def).use { loader ->
            val cls = loader.loadClass(def.name)
            val value: Any = if (kind == "array") arrayOf("text") else "text"
            assertSame(value, cls.getMethod("call", receiver, Any::class.java)
                .invoke(cls.getConstructor().newInstance(), receiver.getConstructor().newInstance(), value))
        }
    }

    @Test
    fun nullableIntersectionArrayPreservesNull() {
        val u = TypeDef.variable("U", TypeDef.of(CharSequence::class.java), TypeDef.of(Serializable::class.java))
        val arrays = MethodDef.builder("nullableArrays").addModifiers(Modifier.PUBLIC, Modifier.STATIC).addTypeVariable(u)
            .addParameter("value", u.array().makeNullable()).returns(TypeDef.OBJECT.makeNullable()).build()
        val def = ClassDef.builder("test.NullableIntersectionArray")
            .addMethod(MethodDef.builder("call").returns(TypeDef.OBJECT.makeNullable())
                .build { _, _ -> ClassTypeDef.of(Calls::class.java).invokeStatic(arrays, ExpressionDef.nullValue()).returning() }).build()
        compileMatchingSnapshots(snapshots, def).use { loader ->
            val cls = loader.loadClass(def.name)
            assertNull(cls.getMethod("call").invoke(cls.getConstructor().newInstance()))
        }
    }

    @ParameterizedTest(name = "successive casts of a {0} value keep every bound")
    @ValueSource(strings = ["scalar", "array", "recursive"])
    fun successiveCastsKeepEveryBound(kind: String) {
        val v = TypeDef.variable("V", TypeDef.of(Serializable::class.java))
        val first = MethodDef.builder("serializable").addModifiers(Modifier.PUBLIC, Modifier.STATIC).addTypeVariable(v)
            .addParameter("value", v).returns(TypeDef.OBJECT).build()
        val u = if (kind == "recursive") TypeDef.variable("U", TypeDef.parameterized(ClassTypeDef.of(Comparable::class.java), TypeDef.variable("U")))
            else TypeDef.variable("U", TypeDef.of(CharSequence::class.java), TypeDef.of(Serializable::class.java))
        val parameter = if (kind == "array") u.array() else u
        val second = MethodDef.builder(when (kind) { "array" -> "arrays"; "recursive" -> "comparable"; else -> "intersection" })
            .addModifiers(Modifier.PUBLIC, Modifier.STATIC).addTypeVariable(u).addParameter("value", parameter).returns(TypeDef.OBJECT).build()
        val def = ClassDef.builder("test.SuccessiveCasts$kind")
            .addMethod(MethodDef.builder("call").addParameter("value", TypeDef.OBJECT).returns(TypeDef.OBJECT)
                .build { _, p -> StatementDef.multi(ClassTypeDef.of(Calls::class.java).invokeStatic(first, p[0]),
                    ClassTypeDef.of(Calls::class.java).invokeStatic(second, p[0]).returning()) }).build()
        compileMatchingSnapshots(snapshots, def).use { loader ->
            val cls = loader.loadClass(def.name)
            val value: Any = if (kind == "array") arrayOf("text") else "text"
            assertSame(value, cls.getMethod("call", Any::class.java).invoke(cls.getConstructor().newInstance(), value))
        }
    }

    @ParameterizedTest(name = "smart casts composed through a {0} keep the overload the model selects")
    @ValueSource(strings = ["parameter", "conditional", "switch"])
    fun composedSmartCastsPreserveOverloadSelection(kind: String) {
        val v = TypeDef.variable("V", TypeDef.of(Serializable::class.java))
        val first = MethodDef.builder("serializable").addModifiers(Modifier.PUBLIC, Modifier.STATIC).addTypeVariable(v)
            .addParameter("value", v).returns(TypeDef.OBJECT).build()
        val chooseAny = MethodDef.builder("choose").addParameter("value", TypeDef.OBJECT).returns(String::class.java)
            .build { _, _ -> ExpressionDef.constant("object").returning() }
        val chooseSerializable = MethodDef.builder("choose").addParameter("value", Serializable::class.java).returns(String::class.java)
            .build { _, _ -> ExpressionDef.constant("serializable").returning() }
        val def = ClassDef.builder("test.ComposedSmartCast$kind").addMethod(chooseAny).addMethod(chooseSerializable)
            .addMethod(MethodDef.builder("call").addParameter("flag", TypeDef.Primitive.BOOLEAN).addParameter("value", TypeDef.OBJECT)
                .returns(String::class.java).build { self, p ->
                    val argument = when (kind) {
                        "conditional" -> p[0].ifTrue(p[1], p[1])
                        "switch" -> ExpressionDef.constant(1).asExpressionSwitch(TypeDef.OBJECT,
                            mapOf(ExpressionDef.constant(1) to p[1]), p[1])
                        else -> p[1]
                    }
                    StatementDef.multi(ClassTypeDef.of(Calls::class.java).invokeStatic(first, p[1]),
                        self.invoke(chooseAny, argument).returning())
                }).build()
        compileMatchingSnapshots(snapshots, def).use { loader ->
            val cls = loader.loadClass(def.name)
            for (flag in listOf(true, false)) {
                assertEquals("object", cls.getMethod("call", Boolean::class.javaPrimitiveType, Any::class.java)
                    .invoke(cls.getConstructor().newInstance(), flag, "text"))
            }
        }
    }

    class IntersectionTarget<X> {
        fun <U> arrays(value: Array<U>): Any where U : List<X>, U : Serializable = value
    }
    class GenericOverloadedChild : CalleeBoundsRegressionTest.Parent<String>() {
        fun <U : Number> identity(value: U): U = value
    }
    open class ArrayParent<X : CharSequence> {
        fun <U : X> arrays(value: Array<U>): Any = value
    }
    class ArrayChild : ArrayParent<String>()
    class Calls {
        companion object {
            @JvmStatic fun <U> nullableArrays(value: Array<U>?): Any? where U : CharSequence, U : Serializable = value
            @JvmStatic fun <U> arrays(value: Array<U>): Any where U : CharSequence, U : Serializable = value
            @JvmStatic fun <U> intersection(value: U): Any where U : CharSequence, U : Serializable = value
            @JvmStatic fun <U : Comparable<U>> comparable(value: U): Any = value
            @JvmStatic fun <V : Serializable> serializable(value: V): Any = value
        }
    }
}
